recipe/patches/0004-Windows-shared-libs.patch

```diff
@@ -0,0 +1,588 @@
+   // piece of wave that we're trying to extract.
+-  int32 wave_start = int32(start_sample - sample_offset),
++  kaldi::int32 wave_start = kaldi::int32(start_sample - sample_offset),
+         wave_end = wave_start + frame_length;
+ 
+   BaseFloat *window = windows + frame * wlda;
+@@ -299,7 +299,7 @@ __global__ void extract_window_kernel(
+     // be reached for about two frames per utterance, so we don't concern
+     // ourselves excessively with efficiency.
+     for (int s = tidx; s < frame_length; s += blockDim.x) {
+-      int32 s_in_wave = s + wave_start;
++      kaldi::int32 s_in_wave = s + wave_start;
+       while (s_in_wave < 0 || s_in_wave >= wave_dim) {
+         // reflect around the beginning or end of the wave.
+         // e.g. -1 -> 0, -2 -> 1.
+diff --git a/src/cudamatrix/cu-allocator.h b/src/cudamatrix/cu-allocator.h
+index d7d65da..7ccd0fa 100644
+--- a/src/cudamatrix/cu-allocator.h
++++ b/src/cudamatrix/cu-allocator.h
+@@ -38,6 +38,7 @@
+ #include "base/kaldi-common.h"
+ #include "util/stl-utils.h"
+ #include "itf/options-itf.h"
++#include "base/kaldi-dll-exports.h"
+ 
+ namespace kaldi {
+ 
+@@ -82,7 +83,7 @@ struct CuAllocatorOptions {
+   }
+ };
+ 
+-extern CuAllocatorOptions g_allocator_options;
++extern CuAllocatorOptions KALDI_CUMATRIX_DLL g_allocator_options;
+ 
+ inline void RegisterCuAllocatorOptions(OptionsItf *po) {
+   g_allocator_options.Register(po);
+@@ -356,7 +357,7 @@ class CuMemoryAllocator {
+ // output to them the free memory and the total memory of the device.
+ std::string GetFreeGpuMemory(int64* free, int64* total);
+ 
+-extern CuMemoryAllocator g_cuda_allocator;
++extern CuMemoryAllocator KALDI_CUMATRIX_DLL g_cuda_allocator;
+ 
+ }  // namespace kaldi
+ 
+diff --git a/src/cudamatrix/cu-device.cc b/src/cudamatrix/cu-device.cc
+index 39bcf37..ac4d20e 100644
+--- a/src/cudamatrix/cu-device.cc
++++ b/src/cudamatrix/cu-device.cc
+@@ -620,7 +620,10 @@ CuDevice::~CuDevice() {
+ 
+ // Each thread has its own copy of the CuDevice object.
+ // Note: this was declared "static".
+-thread_local CuDevice CuDevice::this_thread_device_;
++CuDevice& CuDevice::this_thread_device() {
++    static thread_local CuDevice this_thread_device_;
++    return this_thread_device_;
++}
+ 
+ CuDevice::CuDeviceOptions CuDevice::device_options_;
+ 
+diff --git a/src/cudamatrix/cu-device.h b/src/cudamatrix/cu-device.h
+index 2f278eb..21cb176 100644
+--- a/src/cudamatrix/cu-device.h
++++ b/src/cudamatrix/cu-device.h
+@@ -48,6 +48,7 @@ typedef int cusolverStatus_t;
+ #include "base/timer.h"
+ #include "cudamatrix/cu-allocator.h"
+ #include "cudamatrix/cu-common.h"
++#include "base/kaldi-dll-exports.h"
+ 
+ 
+ namespace kaldi {
+@@ -87,7 +88,7 @@ class CuDevice {
+   // At the beginning of the program, if you want to use a GPU, you
+   // should call CuDevice::Instantiate().SelectGpuId(..).
+   static CuDevice& Instantiate() {
+-    CuDevice &ans = this_thread_device_;
++    CuDevice &ans = this_thread_device();
```

Comment:
Ah, it seems you picked up a few tricks from the gprc work? ;-)

Or if you found out how to deal with `thread_local` vs. `__declspec(dllexport)` independently, I doff my hat!

Reply:
Yes, I meant to mention! I noticed you had a workaround in a follow up commit, so used that here, no hat doffing required.

Reply:
I didn't come up with that workaround myself [either](https://github.com/h-vetinari/grpc/commit/bcb464fe316318903c7744a8dc7dec9a4443c409), I had been completely stuck (hence I'd have been impressed if you had come up with it by yourself 🙃)